src/markdown/tutorial/part-2/11-ember-data.md

```diff
@@ -251,60 +257,97 @@ Darn, there were a couple of failing tests! At the same time, it's great that we
+- Our resource URLs have a `.json` extension at the end.
 
-As it turns out, JSON:API just happens to be EmberData's default data protocol and interchange format. Out of the box, EmberData provides a default JSON:API adapter and serializer. This is great news for us, since that is also what our server has implemented. What a wonderful coincidence!
+The first thing we want to do is have our builder respect a configurable default host and/or namespace. Adding a namespace prefix happens to be pretty common across Ember apps, so EmberData provides a global config mechanism for host and namespace. Typically you will want to do this either in your store file or app file.
 
-However, as mentioned above, there are some minor differences between how our server works and EmberData's default assumptions. We can customize the default behavior by defining our own adapter and serializer:
+```run:file:patch lang=js cwd=super-rentals filename=app/app.js
+@@ -2,6 +2,11 @@
+ import Application from '@ember/application';
+ import Resolver from 'ember-resolver';
+ import loadInitializers from 'ember-load-initializers';
+ import config from 'super-rentals/config/environment';
++import { setBuildURLConfig } from '@ember-data/request-utils';
++
++setBuildURLConfig({
++  namespace: 'api',
++});
 
-```run:file:create lang=js cwd=super-rentals filename=app/adapters/application.js
-import JSONAPIAdapter from '@ember-data/adapter/json-api';
+ export default class App extends Application {
+   modulePrefix = config.modulePrefix;
+```
 
-export default class ApplicationAdapter extends JSONAPIAdapter {
-  namespace = 'api';
+Adding the `.json` extension is a bit less common, and doesn't have a declarative configuration API of its own. We could just modify request options directly in place of use, but that would be a bit messy. Instead, let's create a handler to do this for us.
 
-  buildURL(...args) {
-    return `${super.buildURL(...args)}.json`;
-  }
-}
-```
+```run:file:create cwd=super-rentals filename=app/utils/handlers.js
+const JsonSuffixHandler = {
+  request(context, next) {
+    const { request } = context;
+    const updatedRequest = Object.assign({}, request, {
+      url: request.url + '.json',
+    });
 
-```run:file:create lang=js cwd=super-rentals filename=app/serializers/application.js
-import JSONAPISerializer from '@ember-data/serializer/json-api';
+    return next(updatedRequest);
+  },
+};
 
-export default class ApplicationSerializer extends JSONAPISerializer {}
+export { JsonSuffixHandler };
 ```
 
-By convention, adapters are located at `app/adapters`. Furthermore, the adapter named `application` is called the *application adapter*, which will be used to fetch data for all models in our app.
+As you can see, the handler appends `.json` to the URL of each request. Pretty simple, right? Then it calls the `next` function with the modified request object. This is how we can chain multiple handlers together to build up a request.
```

Comment:
Should this explain why `Object.assign` is used instead of mutating the request object?

That isn't obvious why from this explanation, and makes the code a bit more complicated so I wouldn't assumy it's pretty simple for everyone 😬

Reply:
yeah request suppose to be immutable, so we need to create a new object. Not sure if we need big explanation for why it is immutable, might just add a code comment about it above `Object.assign`

Reply:
Explaining immutability would take this too far indeed. A short one-line reference that the "request is immutable" would be helpful!